### PR TITLE
Omit empty sections in map file output

### LINKED ISSUE #1169: Remove empty sections from text map file
The current format of the text map file is below :

```
Linker Script and memory map
Build Statistics
# <file> <start time> <duration>
Linker scripts used (including INCLUDE command)
Linker Plugin Information
Linker Plugin Run Information

```

These sections are usually empty if  you just link a simple object file

`ld.eld a.o -Map x`

We should remove the empty sections in the map file if there is no content.

### PR DESCRIPTION
**Pull request by Steven6798**
Remove unused sections from the generated text map file when they contain no data. This change improves readability by only including sections with actual content.

Fix #1169

### CODE REVIEW

**Comment by Steven6798** [comment]
I also updated the BuildAndReproduce test to no longer check for the empty "Linker Plugin Information" section

**Review by quic-areg — APPROVED** [review]

File: lib/LayoutMap/TextLayoutPrinter.cpp [hunk 1]
```diff
@@ -357,6 +357,13 @@ void TextLayoutPrinter::printGlobalPluginInfo(Module &M, bool UseColor) {
       M.getScript().getPluginForType(plugin::Plugin::Type::ControlFileSize);
   const LinkerScript::PluginVectorT &UniversalPlugins =
       M.getPluginManager().getUniversalPlugins();
+  bool hasPluginInfo =
+      !UniversalPlugins.empty() || !PluginListForSectionMatcher.empty() ||
+      !PluginListForSectionIterator.empty() ||
+      !PluginListForOutputSectionIterator.empty() ||
+      !PluginListForMemorySize.empty() || !PluginListForFileSize.empty();
+  if (!hasPluginInfo)
+    return;
```
Comment (quic-seaswara) [inline lib/LayoutMap/TextLayoutPrinter.cpp:367]: nit : Can we make this a private function ?

File: lib/LayoutMap/TextLayoutPrinter.cpp [hunk 2]
```diff
@@ -1141,10 +1157,13 @@ void TextLayoutPrinter::printBuildStatistics(Module &CurModule, bool UseColor) {
 
 // Print the loaded Linker scripts and Memory Map files.
 void TextLayoutPrinter::printInputActions(bool UseColor) {
+  auto &InputActions = ThisLayoutInfo->getInputActions();
+  if (InputActions.empty())
+    return;
```
Comment (quic-seaswara) [inline lib/LayoutMap/TextLayoutPrinter.cpp]: Do we have a test that checks this ?

File: lib/LayoutMap/TextLayoutPrinter.cpp (hunk 1 above)
Comment (Steven6798) [reply lib/LayoutMap/TextLayoutPrinter.cpp:367]: Of course.

File: lib/LayoutMap/TextLayoutPrinter.cpp (hunk 2 above)
Comment (Steven6798) [reply lib/LayoutMap/TextLayoutPrinter.cpp]: I forgot to add the tests. Thanks!

**Comment by parth-07** [comment]
@quic-seaswara @quic-areg 
I feel that in the map-file seeing the sections actually empty and the stats as actually `0` is more assuring than not seeing the section/link-stat at all. Users, if they do not see the section/link-stat, may wonder is the stat `0` or did the linker not print the stat, or perhaps this stat is not available in this link version / link configuration. Please share your thoughts on it.

File: lib/LayoutMap/TextLayoutPrinter.cpp (hunk 1 above)
Comment (quic-areg) [reply lib/LayoutMap/TextLayoutPrinter.cpp:367]: I don't think it is valuable to extract this to a function. It is only ever used once and is more readable as-is imo.

File: lib/LayoutMap/TextLayoutPrinter.cpp [hunk 3]
```diff
@@ -1114,8 +1121,15 @@ void TextLayoutPrinter::printLinkerInsertedTimingStats(Module &CurModule) {
...
   if (UseColor)
     outputStream().changeColor(llvm::raw_ostream::MAGENTA);
-  outputStream() << "\nBuild Statistics";
-  outputStream() << "\n# <file> <start time> <duration>\n";
+  bool InsertedBuildStats = false;
+  auto PrintBuildStatsHeader = [this, &InsertedBuildStats]() {
+    if (InsertedBuildStats)
+      return;
+    outputStream() << "\nBuild Statistics";
+    outputStream() << "\n# <file> <start time> <duration>\n";
```
Comment (quic-areg) [inline lib/LayoutMap/TextLayoutPrinter.cpp]: In practice I don't think we ever show this information. It was a hexagon-specific feature (timing section) that I do not think is in use. If that is the case, I think it should be removed entirely. @quic-seaswara

Comment (quic-seaswara) [reply lib/LayoutMap/TextLayoutPrinter.cpp]: Good suggestion, let us remove emitting Build Statistics entirely then. Please raise a seperate PR for this removal.

File: lib/LayoutMap/TextLayoutPrinter.cpp (hunk 2 above)
Comment (Steven6798) [reply lib/LayoutMap/TextLayoutPrinter.cpp]: For this case I don't see a way how this section will ever be empty. We always have at least 1 entry for the input file, and we can't run without input files. Any thoughts?

Comment (quic-seaswara) [reply lib/LayoutMap/TextLayoutPrinter.cpp]: Yes, InputActions will never be empty I suppose.

**Comment by Steven6798** [comment]
I removed the changes from printInputActions function since it is never empty. Also, from printBuildStatistics since it is not used and will be removed in another issue/PR.

File: lib/LayoutMap/TextLayoutPrinter.cpp [hunk 4]
```diff
@@ -358,6 +358,14 @@ void TextLayoutPrinter::printGlobalPluginInfo(Module &M, bool UseColor) {
   const LinkerScript::PluginVectorT &UniversalPlugins =
       M.getPluginManager().getUniversalPlugins();
 
+  bool hasPluginInfo =
+      !UniversalPlugins.empty() || !PluginListForSectionMatcher.empty() ||
+      !PluginListForSectionIterator.empty() ||
+      !PluginListForOutputSectionIterator.empty() ||
+      !PluginListForMemorySize.empty() || !PluginListForFileSize.empty();
+  if (!hasPluginInfo)
+    return;
```
Comment (quic-seaswara) [inline lib/LayoutMap/TextLayoutPrinter.cpp:367]: Can we simply use M.getLinkerScript().getPluginRunList().size() and see if it is non zero, then we print the plugin information ?

Comment (Steven6798) [reply lib/LayoutMap/TextLayoutPrinter.cpp:367]: I'll try that and let you know. Thanks!

Comment (Steven6798) [reply lib/LayoutMap/TextLayoutPrinter.cpp:367]: That's not enough. The EmitMapFileEvenInCrash test fails it is expecting plugin information that is now missing. It looks like we need to check for both plugin information and plugin run information, so I need to add your suggestion at the end of the function to have both checks. Nice catch.
